### PR TITLE
Logger: log polygon fence points

### LINKED ISSUE #20213: Fence/Log: log polygon and circular fences
For support purposes it is good to know where the fences are (polygon, circular, inclusion/exclusion) so we should log these somehow.  Perhaps when the log file is first started.

This comes from this discussion https://discuss.ardupilot.org/t/no-polygone-fences-in-the-dataflash-logs/82566/2

### PR DESCRIPTION
**Pull request by IamPete1**
Currently there is no record of any loaded fence in the log. This adds a message writer like those for rally and mission. 

Fixes #20213

### CODE REVIEW

File: libraries/AP_Logger/LoggerMessageWriter.h [hunk 1]
```diff
@@ -73,6 +73,25 @@ class LoggerMessageWriter_WriteAllRallyPoints : public LoggerMessageWriter {
...
 
+#if AC_FENCE
+class LoggerMessageWriter_Write_Polyfence : public LoggerMessageWriter {
+public:
+
+    void reset() override;
+    void process() override;
+
+private:
+    enum Stage {
```
Comment (peterbarker) [inline libraries/AP_Logger/LoggerMessageWriter.h:84]:
```suggestion
    enum class Stage {
```

will show up a bug,  I think....

File: libraries/AP_Logger/LoggerMessageWriter.cpp [hunk 2]
```diff
@@ -199,6 +211,19 @@ bool LoggerMessageWriter_DFLogStart::writeallrallypoints()
 }
 #endif
 
+#if AC_FENCE
+bool LoggerMessageWriter_DFLogStart::writeallfence()
+{
+    if (stage != Stage::DONE) {
+        return false;
+    }
+    stage = Stage::RUNNING_SUBWRITERS;
```
Comment (peterbarker) [inline libraries/AP_Logger/LoggerMessageWriter.cpp:220]: This is weird - this isn't part of that enum.  Looks like we need to sprinkle some `enum class`s around.

**Review by peterbarker — DISMISSED** [review]

File: libraries/AP_Logger/LoggerMessageWriter.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_Logger/LoggerMessageWriter.cpp:222]: Its not the same stage enum as the one added in this PR. I classed them all in anycase.

File: libraries/AP_Logger/LoggerMessageWriter.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_Logger/LoggerMessageWriter.h]: done

**Comment by IamPete1** [comment]
Rebased, but not yet re-tested.

**Comment by IamPete1** [comment]
Re-tested.
image

**Review by rmackay9 — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Comment by Williangalvani** [comment]
support is now live in http://plotbeta.ardupilot.org 
image

**Pull request merged**